### PR TITLE
NCDHC mapping

### PR DESCRIPTION
**Pull request by guegueng**
second attempt at NCDHC mapping. Someone should pay particular attention to the date mapping as I’m not confident it’s done correctly.

### CODE REVIEW

File: heidrun/ncdhc.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+    creator :class => DPLA::MAP::Agent, 
+    		:each => record.field('mods:name')
+    		.select { |name| name['mods:role'].map(&:value).include?('creator') }, 
+    		:as => :creator_role do
+      providedLabel creator_role.field('mods:namePart')
+    end
+
+    date :class => DPLA::MAP::TimeSpan,
```
Comment (anarchivist) [inline heidrun/ncdhc.rb]: From the crosswalk, the date mapping is as follows:

> Within `<originInfo>`, the FIRST `<dateCreated keyDate="yes">` and LAST `<dateCreated keyDate="yes">`.

(this was just added for reference)

File: heidrun/ncdhc.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    		.select { |name| name['mods:role'].map(&:value).include?('creator') }, 
+    		:as => :creator_role do
+      providedLabel creator_role.field('mods:namePart')
+    end
+
+    date :class => DPLA::MAP::TimeSpan, 
+    		 :each => record.field('mods:originInfo'), 
+    		 :as => :created do
+      providedLabel created.field('mods:dateCreated').match_attribute(:keyDate, 'yes')
+      self.begin created.field('mods:dateCreated').first_value
```
Comment (anarchivist) [inline heidrun/ncdhc.rb]: Should this be the following?

``` ruby
self.begin created.field('mods:dateCreated').match_attribute(:keyDate, 'yes').first_value
```

File: heidrun/ncdhc.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      self.end created.field('mods:dateCreated').last_value
```
Comment (anarchivist) [inline heidrun/ncdhc.rb]: Should this be the following?

``` ruby
self.end created.field('mods:dateCreated').match_attribute(:keyDate, 'yes').last_value
```

File: heidrun/ncdhc.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    identifier record.field('mods:identifier')
+
+    language :class => DPLA::MAP::Controlled::Language, 
+    				 :each => record.field('mods:language', 'mods:languageTerm'), 
+    				 :as => :lang do
+      providedLabel lang
+    end
+
+    spatial :class => DPLA::MAP::Place, 
+    	      :each => record.field('mods:subject'),
```
Comment (anarchivist) [inline heidrun/ncdhc.rb]: Will this lead to empty instances of `DPLA::MAP::Place` when there are instances of `<mods:subject/>` that _don't_ contain `<mods:geographic/>`? cc @no-reply

File: heidrun/ncdhc.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      providedLabel place.field('mods:geographic')
+    end
+
+    publisher :class => DPLA::MAP::Agent, 
+    					:each => record.field('mods:originInfo'), 
+    					:as => :publisher do
+      providedLabel publisher.field('mods:publisher')
+    end
+    
+    relation record.field('mods:reltedItem')
```
Comment (anarchivist) [inline heidrun/ncdhc.rb]: typo - should be `relation record.field('mods:relatedItem')`

Comment (anarchivist) [inline heidrun/ncdhc.rb]: I think this and the next few lines should be:

``` ruby
relation record.field('mods:relatedItem').fields(['mods:location', 'mods:url'], ['mods:titleInfo', 'mods:title'])
```

cc @no-reply

File: heidrun/ncdhc.rb (hunk 4 above)
Comment (no-reply) [inline heidrun/ncdhc.rb]: That seems right, @anarchivist. I think the best way to handle this example is to do `:each => record.field('mods:subject', 'mods:geographic')`. 

If we needed other values of the subject, we could do `:each => record.field('mods:subject').select { |s| s.child?('mods:geographic') }` (maybe we need a tighter syntax for this? `record.field('mods:subject').with_subfield('mods:geographic')`), but it doesn't seem like it's needed here.

File: heidrun/ncdhc.rb (hunk 5 above)
Comment (no-reply) [inline heidrun/ncdhc.rb]: I'm not sure what we're aiming to map here.

The existing mapping will only select cases where there's a path `mods:location/mods:url/mods:titleInfo/mods:title`, which seems wrong. 

@anarchivist's example should create `relation`s for _both_ the `mods:location/mods:url` and the `mods:titleInfo/mods:title`. So a single `mods:relatedItem` will have multiple `dc:relation`s in the mapped record.

**Comment by no-reply** [comment]
I was able to run this against a batch of MODS records. I think we're ready to get NC into QA.

**Pull request merged**